apps/desktop/src/components/SearchIndexContext.tsx

```diff
@@ -0,0 +1,119 @@
+import { createContext, useContext, useState, ReactNode, useCallback } from 'react';
+import { readNote } from '../commands';
+import { FileEntry } from '../types';
+
+export interface NoteIndexEntry {
+  path: string;
+  title: string;
+  content: string;
+}
+
+interface SearchIndexContextProps {
+  buildIndex: (files: FileEntry[]) => Promise<void>;
+  updateEntry: (path: string, content: string) => void;
+  search: (query: string) => NoteIndexEntry[];
+  isIndexing: boolean;
+}
+
+const SearchIndexContext = createContext<SearchIndexContextProps | undefined>(undefined);
+
+export const useSearchIndex = () => {
+  const context = useContext(SearchIndexContext);
+  if (!context) {
+    throw new Error('useSearchIndex must be used within a SearchIndexProvider');
+  }
+  return context;
+};
+
+export const SearchIndexProvider = ({ children }: { children: ReactNode }) => {
+  const [index, setIndex] = useState<Map<string, NoteIndexEntry>>(new Map());
+  const [isIndexing, setIsIndexing] = useState(false);
+
+  const parseTitle = (path: string, content: string): string => {
+    const lines = content.split('\n');
+    for (const line of lines) {
+      const trimmed = line.trim();
+      if (trimmed === '') continue;
+
+      if (trimmed.startsWith('# ')) {
+        return trimmed.substring(2).trim();
+      } else {
+        // First non-empty line is not a heading
+        break;
+      }
+    }
+
+    // Fallback to basename
+    // Handle both / and \ just in case, though standardizing on / is better
+    const normalizedPath = path.replace(/\\/g, '/');
+    const parts = normalizedPath.split('/');
+    const filename = parts[parts.length - 1];
+    return filename.replace(/\.md$/, '');
+  };
+
+  const buildIndex = useCallback(async (files: FileEntry[]) => {
+    setIsIndexing(true);
+    const newIndex = new Map<string, NoteIndexEntry>();
+    const mdFiles = files.filter(f => !f.is_dir && f.path.endsWith('.md'));
+
+    // We process sequentially to avoid overwhelming the backend,
+    // though Promise.all could be faster if the backend handles concurrency well.
+    // Given it's a local FS, sequential is safer for now.
```

Comment:
Does the async lib handle queuing?

Reply:
Out of curiosity, how does the backend handle the concurrency‽ Can you give an explanation?